### PR TITLE
feat: add POST end point to mark calls as no-shows

### LINKED ISSUE #15686: [CAL-4000] POST end point to mark calls as no-shows
### Is your proposal related to a problem?

Requested here:- [https://calcominc.slack.com/archives/C06UTF7QLU9/p1720211293270259](https://calcominc.slack.com/archives/C06UTF7QLU9/p1720211293270259)

Related PRs [https://github.com/calcom/cal.com/pull/15502/files](https://github.com/calcom/cal.com/pull/15502/files)

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-4000](https://linear.app/calcom/issue/CAL-4000/post-end-point-to-mark-calls-as-no-shows)</sub>

### PR DESCRIPTION
**Pull request by Udit-takkar**
## What does this PR do?

- Fixes https://github.com/calcom/cal.com/issues/15686
- Fixes CAL-4000 (Linear issue number - should be visible at the bottom of the GitHub issue description)

Screenshot 2024-07-08 at 10 07 33 PM

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [x] N/A I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

Make a request to /api/v2/bookings/{bookingUid}/mark-no-show

```
{
    "attendees": [
        {
            "email": <EMAIL_OF_ANY_ATTENDEE>,
            "noShow": true
        }
    ]
}
```

### CODE REVIEW

**Comment by keithwillcode** [comment]
@Udit-takkar why is this being added to v1 and not v2?

**Comment by Udit-takkar** [comment]
> @Udit-takkar why is this being added to v1 and not v2?

Because the customer that requested this endpoint is using v1 API.

**Review by keithwillcode — DISMISSED** [review]
This should be moved to v2. We are no longer adding new functionality to v1. Also, `no-show` should not be the route in the URL. It doesn't point to any resource on the server. Instead, we should go through the bookings route.

File: apps/api/v2/src/ee/bookings/controllers/bookings.controller.ts
Comment (Udit-takkar) [inline apps/api/v2/src/ee/bookings/controllers/bookings.controller.ts:50]: TODO: new version

File: packages/features/handleMarkNoShow.ts
Comment (Udit-takkar) [inline packages/features/handleMarkNoShow.ts:27]: no change in the code here.

File: apps/api/v2/src/ee/bookings/controllers/bookings.controller.ts
Comment (hariombalhara) [inline apps/api/v2/src/ee/bookings/controllers/bookings.controller.ts:249]: I think `/:bookingUid/no-show` makes more sense. As we are making an update to the Booking resource and not all Bookings resource

Comment (Udit-takkar) [reply apps/api/v2/src/ee/bookings/controllers/bookings.controller.ts:249]: This endpoint also updates `Attendee`  table noShow field.

File: apps/api/v2/src/ee/bookings/controllers/bookings.controller.ts
Comment (hariombalhara) [inline apps/api/v2/src/ee/bookings/controllers/bookings.controller.ts:251]: Completely unaware of it, so I want to know why we are using ApiAuthGuard for some endpoints and not for others within the controller

File: apps/api/v2/src/ee/bookings/controllers/bookings.controller.ts
Comment (hariombalhara) [reply apps/api/v2/src/ee/bookings/controllers/bookings.controller.ts:249]: Yeah, I still think it is better because Attendee is connected to booking.


Though there could be a separate endpoint in future to work with attendee directly if required, that would also allow no-show update.

File: packages/features/handleMarkNoShow.ts
Comment (Udit-takkar) [reply packages/features/handleMarkNoShow.ts:90]: @hariombalhara zomars added this code because constructor returns this async function.

```
  constructor(options: Parameters<typeof getWebhooks>[0]) {
    return (async (): Promise<WebhookService> => {
      this.options = options;
      this.webhooks = await getWebhooks(options);
      return this;
    })() as unknown as WebhookService;
  }
  ```

Comment (hariombalhara) [reply packages/features/handleMarkNoShow.ts:90]: Wow, that's the first time I am seeing async constructor. I am not sure if it is working as per expectation but I have never seen constructor doing async work.

Also, eslint is warning about it. But as it is existing code I am fine with it for now.

File: packages/features/handleMarkNoShow.ts
Comment (hariombalhara) [inline packages/features/handleMarkNoShow.ts:107]: We are updating either attendees `noShow` or `hostNoShow` in an API call. Through frontend I think it was perfectly fine as user takes action one by one but through API, I believe developers would want it to be done in same call soon.

File: apps/api/v2/src/ee/bookings/controllers/bookings.controller.ts
Comment (keithwillcode) [reply apps/api/v2/src/ee/bookings/controllers/bookings.controller.ts]: Yeah agreed with @hariombalhara. You shouldn’t be able to directly update attendees without going through the booking

File: apps/api/v2/src/ee/bookings/controllers/bookings.controller.ts
Comment (hariombalhara) [inline apps/api/v2/src/ee/bookings/controllers/bookings.controller.ts:317]: The message "Error while creating no-show booking" won't make sense.

File: apps/api/v2/src/ee/bookings/outputs/mark-no-show.output.ts
Comment (hariombalhara) [inline apps/api/v2/src/ee/bookings/outputs/mark-no-show.output.ts:17]: `message` is for frontend. I think it won't make sense for API. It is translation key anyway.

image

**Review by hariombalhara — DISMISSED** [review]
Tested. Looks good functionality wise !! 

I have some suggestions but as the PR is a bit pressing, they can be addressed later.

Approving it but we should wait for someone from @calcom/platform to review it.

**Review by ThyMinimalDev — CHANGES_REQUESTED** [review]
Looks good functionality wise, I would take into consideration @hariombalhara comments before merging

File: apps/api/v2/src/ee/bookings/controllers/bookings.controller.ts
Comment (ThyMinimalDev) [inline apps/api/v2/src/ee/bookings/controllers/bookings.controller.ts:200]: we should try to use a verb in the URL instead of no-show

**Review by ThyMinimalDev — APPROVED** [review]

**Pull request merged**
